Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,589 @@
+-- |
+-- Module:      Math.NumberTheory.DirichletCharacters
+-- Copyright:   (c) 2018 Bhavik Mehta
+-- Licence:     MIT
+-- Maintainer:  Bhavik Mehta <bhavikmehta8@gmail.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- Implementation and enumeration of Dirichlet characters.
+--
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.DirichletCharacters
+  (
+  -- * Roots of unity
+    RootOfUnity
+  -- ** Conversions
+  , toRootOfUnity
+  , fromRootOfUnity
+  , toComplex
+  -- * An absorbing semigroup
+  , OrZero, pattern Zero, pattern NonZero
+  , asNumber
+  -- * Dirichlet characters
+  , DirichletCharacter
+  -- ** Construction
+  , indexToChar
+  , indicesToChars
+  , characterNumber
+  , allChars
+  -- ** Evaluation
+  , evaluate
+  , generalEval
+  , toFunction
+  , allEval
+  -- ** Special Dirichlet characters
+  , principalChar
+  , isPrincipal
```

Comment:
Isn't it redundant to export `isPrincipal`, when we have `Eq` instance already available? I would rather keep the export list as laconic as practical.

Reply:
Not really. I think your idea is that `isPrincipal chi` should be equivalent to `chi == mempty`, but `mempty` requires a `KnownNat n` instance, while `isPrincipal` does not. 
(`mempty` should require this instance, since to create a character from thin air, we need to know the modulus on term level)

Reply:
Make sense, thanks.